gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,429 @@
+
+        data = await self.redis.get(key)
+
+        return SessionInfo.model_validate_json(data)
+
+    async def delete_qod_session(self, id: str) -> SessionInfo:
+        key = f"{_prefix_gateway_nef}:{id}"
+        nef_id = await self.redis.get(key)
+
+        sub_info = await self.get_qod_information_by_id(id)
+
+        if sub_info.qosStatus == Status.UNAVAILABLE:
+            return sub_info
+
+        res = await self.httpx_client.delete(nef_id)
+
+        if res.status_code == 404:
+            raise ResourceNotFound()
+
+        if not res.is_success:
+            raise BadRequest()
+
+        if res.status_code == 204:
+            sub_info.qosStatus = Status.UNAVAILABLE
+            sub_info.statusInfo = StatusInfo.DELETE_REQUESTED
+
+        await self._save_subscription_info(id, sub_info)
+
+        return sub_info
+
+    async def get_qod_information_device(self, device: Device) -> List[SessionInfo]:
+        id = []
+
+        if device.phoneNumber:
+            key = f"{_prefix_device}:{device.phoneNumber}"
+            id = await self._get_device(key)
+        elif device.networkAccessIdentifier:
+            key = f"{_prefix_device}:{device.networkAccessIdentifier}"
+            id = await self._get_device(key)
+        elif device.ipv4Address and device.ipv4Address.privateAddress:
+            key = f"{_prefix_device}:{device.ipv4Address.privateAddress}"
+            id = await self._get_device(key)
+        elif device.ipv4Address and device.ipv4Address.publicPort:
+            key = f"{_prefix_device}:{device.ipv4Address.publicAddress}:{device.ipv4Address.publicPort}"
+            id = await self._get_device(key)
+        elif device.ipv6Address:
+            key = f"{_prefix_device}:{device.ipv6Address.exploded}"
+            id = await self._get_device(key)
+
+        if len(id) == 0 or id is None:
+            raise ResourceNotFound()
+
+        sub_info = []
+
+        for sub_id in id:
+            sub_info.append(await self.get_qod_information_by_id(sub_id))
+
+        return sub_info
+
+    async def extend_session(self, id: str, req: ExtendSessionDuration) -> SessionInfo:
+        sub_info = await self.get_qod_information_by_id(id)
+
+        if sub_info.device is None:
+            raise ResourceNotFound()
+
+        key = f"{_prefix_gateway_nef}:{id}"
+        nef_id = await self.redis.get(key)
+
+        payload = AsSessionWithQoSSubscription(
+            self=nef_id,
+            supportedFeatures="800820",
+            notificationDestination=Link(
+                f"{self.notification_url}/callbacks/v1/qod/{id}"
+            ),
+            flowInfo=self._get_flow_description(sub_info),
+            qosReference=sub_info.qosProfile,
+            usageThreshold=UsageThreshold(
+                duration=sub_info.duration + req.requestedAdditionalDuration,
+                totalVolume=0,
+                downlinkVolume=0,
+                uplinkVolume=0,
+            ),
+        )
```

Comment:
The objective of using PATCH, was so that you didn't have to fill everything again. Here from what I'm seeing only the usage thresholds are changing